### ISSUE TITLE
Request fails with 400 for empty Host header

### ISSUE THREAD

**Issue by saxxi** [ninenines/cowboy#1294]
It appears AWS Application Load Balancer's health check ping is not sending Host header which causes the machine to wrongly think it's unhealthy and therefore terminate it.

```
$ curl -i localhost:4000/status  -H 'Host:anything-here-is-okay'
HTTP/1.1 200 OK
server: Cowboy
date: Mon, 04 Jun 2018 07:18:34 GMT
content-length: 19
cache-control: max-age=0, private, must-revalidate

{"version":"0.0.3"}%
```

And this is what we think is happening on the server side, ie from AWS ALB's health check, reproducible in local env too:

```
$ curl -i localhost:4000/status  -H 'Host:'
HTTP/1.1 400 Bad Request
connection: close
server: Cowboy
date: Mon, 04 Jun 2018 07:19:17 GMT
content-length: 0
```

The logged error (same as production):

```
[error] Cowboy returned 400 because it was unable to parse the request headers.

This may happen because there are no headers, or there are too many headers
or the header name or value are too large (such as a large cookie).

You can customize those values when configuring your http/https
server. The configuration option and default values are shown below:

    protocol_options: [
      max_header_name_length: 64,
      max_header_value_length: 4096,
      max_headers: 100,
      max_request_line_length: 8096
    ]
```

Regarding the solution to the issue I was hoping we could add or replace a plug before this line kicks in and inject something like:

```
# Request -> [OurMiddleware] -> [Plug] -> [Phoenix] -> ...

defmodule NotificationServiceWeb.Endpoint do
  plug :add_host_if_blank
  use Phoenix.Endpoint

  defp add_host_if_blank
      conn = conn
      |> put_req_header("Host", "custom-host") # this seems to work only on testing?
  end
end
```

I posted this issue on [phoenix's group](https://github.com/phoenixframework/phoenix/issues/2437) however it seems the issue comes from `Cowboy`. Any thoughts on how to fix this?

**Comment by essen** [comment]
Sending a request without a host header is an error, sending a request with an empty host header is not. The test for an empty host header has not been written yet so it's possible it's malfunctioning. I'll add this test later today.

**Comment by essen** [comment]
I've written a test and see Cowboy is behaving as intended. Are you sure Cowboy is configured properly? Cowboy sends a 400 when there's no entry in the routes corresponding to the empty host. If you use `'_'` in the routes or even `""` it works as intended.

**Comment by saxxi** [comment]
I've created a really simple app, nothing fancy really and still able to reproduce the error.

https://github.com/saxxi/cowboy-elixir-empty-headers-error

**Comment by essen** [comment]
You'll need to provide an Erlang sample app for me to be able to read it, or at the very least a smaller Elixir app, I'm not setting up postgres for this.

Like I've said though a 400 is the expected result when the host is not known, can you check that this isn't your issue? Just use `ranch:get_protocol_options(Name)` where `Name` is the name of the listener.

**Comment by saxxi** [comment]
I've updated the test application to remove the ecto dependency but this really how I got my app running:

```
mix phx.new app_tester --no-brunch --no-html
```

And I changed `router.ex` and `status_controller.ex`, nothing else.

Unfortunately I see you're trying to point me to look into `ranch:get_protocol_options`, however due to little knowledge in elixir/erlang I'm still trying to figure out how should I use it.

**Comment by essen** [comment]
Please give the full list of commands I'm supposed to run at least.

**Comment by saxxi** [comment]
I've update the repo's `README`:

```
git clone git@github.com:saxxi/cowboy-elixir-empty-headers-error.git
mix deps.get
iex -S mix phx.server
```

Thanks for staying by.

**Comment by essen** [comment]
Oh sorry I missed that. This is Cowboy 1.x. I would recommend upgrading to Cowboy 2 instead. I am no longer maintaining the 1.x version.

That said, you can *probably* use a Cowboy `onrequest` hook to edit the request using `cowboy_req:get` and `cowboy_req:set`.

**Comment by saxxi** [comment]
No worries.

I've updated [my code](https://github.com/saxxi/cowboy-elixir-empty-headers-error) to use `phoenix v.1.4-dev` and `Cowboy 2.0` and this again reports the same error.

The code is divided in two clear commits explaining what I have done in order to create the project and  what [I added](https://github.com/saxxi/cowboy-elixir-empty-headers-error/commit/d397f4dbc98a95491991af922f9b2707dc713104).

The issue behind this is AWS application load balancer's health check is returning this kind of response as they're probably trying to cut on bandwidth and therefore shrinking the header requests at maximum. We don't know the actual request as it's coming with an error before we can catch it and process it anyhow. This makes the Elastic Container Service think the application is unhealthy and therefore terminate it asap.

Using a TCP load balancer won't also not help us as then we'd be forced to mess with certificates and whatnot.

We even tried to use a static empty file but this also returned an error as this library loads in before anything else.

Any help on how to disable this feature or manually inject some monkey patch or even fork this repo in order to fix those particular lines involved would be appreciated.

Thanks again, A.

**Comment by essen** [comment]
Please update Cowboy to at least 2.4 and if that does not work, try master. If it's fixed only in master I can issue a new release soon.

I do not know when this was fixed and/or where your issue occurs in exactly. You'd have to provide Erlang sample projects otherwise my knowledge is limited.

**Comment by OvermindDL1** [comment]
@essen  Technically what cowboy is doing is correct, the spec itself does not allow for a missing Host header when in HTTP/1.1 mode.  Amazon's heartbeat is very much in the wrong here.  However considering how used aws is it might be good to make it a configurable option to not follow the spec in this case (but default to following it)?

**Comment by essen** [comment]
Which goes back to my initial comment: is it a *missing* host header (which is wrong) or an *empty* host header (which is correct)? I've confirmed the latter is working as intended, but if it's the former I can't really do anything about that.

**Comment by saxxi** [comment]
On the [third commit](https://github.com/saxxi/cowboy-elixir-empty-headers-error/commits/master) I've tried bumping the version to `2.4` and `master`.

I believe this constitute as `empty` Host but I guess can be debatable:

```
curl -i localhost:4000/status  -H 'Host:'
```

While probably not conformed to the RFC this works for example on `Ruby On Rails`.

I've already opened a ticket on AWS as they should really fix this.

**Comment by OvermindDL1** [comment]
> Which goes back to my initial comment: is it a missing host header (which is wrong) or an empty host header (which is correct)? 

It is 'missing', not empty.

https://forums.aws.amazon.com/thread.jspa?messageID=423533

**Comment by essen** [comment]
Yes when you give curl `'Host:'` it acts as if it's missing. If you give `'Host: ` (with a space) it acts as if it's empty. With your example I get a 400 if missing, and 404 if empty (not sure why 404, it's going through Phoenix or something, but that means the request goes through just fine). So from my point of view it's working as intended.

If you want to work around this with Cowboy 2.4 take a look at the function `cowboy_http:request/3`, it checks for the existence of the host header there. Just remove the first clause.

**Comment by justgage** [comment]
For other people, I seem to be getting this error when I use Postman on a Mac. I'm using version `6.7.1`. CURL seemed to be fine. Not sure why it's messing up so badly.

**Comment by essen** [comment]
Maybe worth adding an option to treat a missing host header the same as empty for HTTP/1.1 (effectively making it work the same as HTTP/1.0). Open a new ticket if that's desirable.
